libbpf-tools/execsnoop.bpf.c

```diff
@@ -0,0 +1,142 @@
+#include "vmlinux.h"
+#include <bpf/bpf_helpers.h>
+#include "execsnoop.h"
+
+// null isn't defined
+#define NULL 0
+// from linux/bpf.h
+#define BPF_NOEXIST  1
+
+#define BASE_EVENT_SIZE (sizeof(struct event) - FULL_MAX_ARGS_ARR)
+#define EVENT_SIZE(X) (BASE_EVENT_SIZE + (ARGSIZE * (X)->args_count))
+
+const volatile bool ignore_failed = true;
+const volatile uid_t targ_uid = INVALID_UID;
+const volatile int max_args = DEFAULT_MAXARGS;
```

Comment:
Python version allows this to be adjustable (`--max-args` argument). Although  it's changes [defined value](https://github.com/iovisor/bcc/blob/master/tools/execsnoop.py#L219).

Reply:
alright, then just stick to max_args, no need for TOTAL_MAX_ARGS (as I mentioned, recent enough verifier will know that max_args is a constant). Verifier also supports proper loop, so you don't need #pragma unroll either (in fact #pragma unroll won't work with const max_args).

Reply:
It's seems in **5.4** it isn't recent enough, cause without this it's rejects code.

Reply:
> right, I forgot execsnoop doesn't really need any of CO-RE stuff and can run on much older kernels. This tool can be made to run even on kernels that don't have BTF. You'll need to drop vmlinux.h and just copy/paste trace_event_raw_sys_enter/trace_event_raw_sys_exit struct definitions. It will make this tool quite widely runnable. Not sure how important that is, but something to consider for sure. cc @brendangregg

Basically what kernel version should I target, when I was working on this I was targeting latest LTS (which is `5.4`), but you mentioned that I could target even older kernel version, so question is should I ? Or maybe I should target something newer than `5.4` (because it had some issues with verifier).

Reply:
@bacher09, it seems like you ignored most of the important details I showed in https://github.com/anakryiko/bcc/commit/cda6c3411787504bc7c1dcf8af419a6e86b923c3. Wasn't that demo working for 5.4? See size_t types for len, `strs` variable to keep track of end of argument strings, negative arg_cnt value to mark `had_more_args` fact without wasting 4 bytes in the event struct, there is also a nice trick with reusing return result field between kprobe/kretprobe to store actual full length of the data.

So please do check it again and please let me know if that approach still works on 5.4? If yes, just try to follow the spirit of it and it should work for execsnoop here as well. If not, please paste verifier log showing verification failure on 5.4, we'll need to figure out what's wrong. This whole approach should work on all kernels that have BTF built-in. Don't remember if it's 5.4, 5.2 or whichever, we can check later precisely, but I'm sure 5.4 should be supported.

Reply:
`Wasn't that demo working for 5.4?`

Exactly it wasn't, I even wrote a [comment about this](https://github.com/anakryiko/bcc/commit/f001f8cadf69c6beee10476beb34ceee76b284c1#r37837678).

Reply:
Hm.. you must have missed that I actually updated first revision. See https://github.com/anakryiko/bcc/commit/f001f8cadf69c6beee10476beb34ceee76b284c1#r37842380.

Reply:
Regarding `more_args` I added to make code more readable, as I explained in another PR and it actually redundant (cause `more_args` true only when `args_count == max_args`).

Reply:
Doing something like below on user-space side:
```
bool more_args = false;

if (event->args_count < 0) {
    event->args_count = -event->args_count;
    more_args = true;
}
```

isn't readable enough?

As for `args_count == max_args` wouldn't it be the case when actual number of args is exactly equal to max_args?

Reply:
Oops sorry, updated version indeed works on **5.4**, but my version also works on **5.4** if I change `for` loop in enter probe to be unrolled (like this was previously done).

Reply:
So wanted to ask, should I return back `pragma unroll` before for loop to support **5.4** ?

Also wanted to point it out that currently `Makefile` incorrectly tracks dependencies (some headers isn't tracked as dependency). For example if I change `execsnoop.h` and do `make execsnoop` make will output `make: 'execsnoop' is up to date.`, while it isn't correct. And it's same behaviour for all other tools.

And also another question, is any plans to support cmake build, cause I did small sample how  it [could look like](https://gist.github.com/bacher09/beef9131ceece01c6980908a5ca5a560#file-cmakelists-txt-L84) (it just a sample and doesn't handle all the tools). And in that case there would be no need to update `.gitignore` for each tool, in addition it would be possible to make tools that would use multiple bpf sources (if this desirable).

Reply:
Yeah, I'd add `pragma unroll back`.

As for Makefile not tracking depenencies. Yeah, seems like $(wildcard %.h) doesn't record dependency properly. Adding plain %.h will break on tools that don't have corresponding .h header... Not sure, need to find some other way. Thanks for bringing this up.

As for cmake, I have almost 0 experience with cmake, so didn't even attempt to write one. I actually like Makefile for it's simplicity and straightforwardness (once you learn Makefile basics, of course). cmake has too much magic for me. But if someone were to write a good cmake for all libbpf-tools that are easy to extend, that would be great, of course.